src/libsyntax/parse/parser.rs

```diff
@@ -547,40 +547,38 @@ impl<'a> Parser<'a> {
         }
     }
 
-    crate fn check_ident(&mut self) -> bool {
-        if self.token.is_ident() {
+    fn check_or_expected(&mut self, ok: bool, mk_type: impl FnOnce() -> TokenType) -> bool {
```

Comment:
I'd prefer to avoid closure passing unless it's really necessary.
Right now all the actual uses construct `TokenType` in trivial ways, so we can avoid it and just pass `TokenType`.